### ISSUE TITLE
Python 3.6 compatability

### ISSUE THREAD

**Issue by lschmelzeisen** [borntyping/jsonlog#1]
Hi, this seems like a great little tool, that I would like to use for my projects. However, I'm bound to support python 3.6 whereas in the Readme it says that it is written for 3.7 but then the `pyproject.toml` says 3.8 :)

Because you also mentioned that compatibility patches will be accepted, I was wondering if you would be ok to add support for 3.6 and what would need to change for that.

* The main problem with transitioning to support 3.6 are dataclasses (use in `jsonlog.foramtter` and multiple `jsonlog-cli` submodules) which are only available in 3.7 and above. Luckily, there is a [backport of dataclasses to 3.6](https://pypi.org/project/dataclasses/). This could be added to `[tool.poetry.dependencies]` like this: `dataclasses = "*; python_version<3.7"`.

With this (and changing the python version requirements in `pyproject.toml` and `setup.cfg`, I was able to successfully run all tests for `jsonlog` in 3.6. However, for `jsonlog-cli` a few more things need to be done:

* `from __future__ import annotations` is only available in 3.7 and above. To accommodate 3.6 these lines would need to be removed and type annotations for methods that return there own classes would need to be string wrapped. So, for instance `-> ColourMap` would need to change to `-> "ColourMap"`.
* `typing.Protocol` (used in `jsonlog_cli.record` and `jsonlog_cli.stream`) is only available in 3.8 and above. There exists is a backport in [typing-extensions](https://pypi.org/project/typing-extensions/) that could be used here.

This seems all that would need to be changed, however, I could not verify this as currently `jsonlog-cli`'s tests seem to be broken. (By the way, using the current pip-version of `jsonlog-cli` in 3.8, I was not able visualize the provided `docs/example.log` and always received a `json.decoder.JSONDecodeError`.)

In summary, support 3.6 should be relatively easy and I would be willing to submit a pull request for this once `jsonlog-cli` is fixed. Would this be something that you are interested in?

**Comment by borntyping** [comment]
The changes you've described sound great, and I'd be very interested in merging those.

I'll try and fix the tests soon - I'm currently on holiday, so this might not be until next week. Setting up proper CI for the tests is something I've not put time into yet, as it's been a bit of a hobby project with a lot of rapid changes to the API (mostly on the `jsonlog-cli` side).

**Comment by borntyping** [comment]
The README should probably say it's been written targetting 3.8 (I switched during development). I'll probably also consider noting support for earlier versions of Python until their EOL date, but being a bit more careful to do a "final" release that supports those versions (something I wished I had done for my last logging project!).

**Comment by lschmelzeisen** [comment]
Sounds great! Just ping me once you get around to this. Proper CI would of course be very helpful for me when testing my changes :) Enjoy your holidays!

**Comment by borntyping** [comment]
I've fixed up the tests, cleaned up the README, and added CI using GitHub actions - should give you a clean run at this :)

https://github.com/borntyping/jsonlog/actions/runs/135972227

**Comment by lschmelzeisen** [comment]
Unfortunately, today's change to use `pydantic` (to fix #5) broke 3.6 compatibility again:

```
jsonlog --help
Traceback (most recent call last):
  File "/home/lukas/test/.venv/bin/jsonlog", line 7, in <module>
    from jsonlog_cli.cli import main
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/jsonlog_cli/cli.py", line 6, in <module>
    import jsonlog_cli.config
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/jsonlog_cli/config.py", line 11, in <module>
    import jsonlog_cli.pattern
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/jsonlog_cli/pattern.py", line 16, in <module>
    class Pattern(pydantic.BaseModel):
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/main.py", line 251, in __new__
    config=config,
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 319, in infer
    field_info=field_info,
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 272, in __init__
    self.prepare()
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 364, in prepare
    self._type_analysis()
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 465, in _type_analysis
    self.key_field = self._create_sub_type(self.type_.__args__[0], 'key_' + self.name, for_keys=True)
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 494, in _create_sub_type
    model_config=self.model_config,
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 272, in __init__
    self.prepare()
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 364, in prepare
    self._type_analysis()
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 428, in _type_analysis
    self.sub_fields = [self._create_sub_type(t, f'{self.name}_{display_as_type(t)}') for t in types_]
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 428, in <listcomp>
    self.sub_fields = [self._create_sub_type(t, f'{self.name}_{display_as_type(t)}') for t in types_]
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 494, in _create_sub_type
    model_config=self.model_config,
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 272, in __init__
    self.prepare()
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 370, in prepare
    self.populate_validators()
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/fields.py", line 508, in populate_validators
    *(get_validators() if get_validators else list(find_validators(self.type_, self.model_config))),
  File "/home/lukas/test/.venv/lib/python3.6/site-packages/pydantic/validators.py", line 584, in find_validators
    raise RuntimeError(f'no validator found for {type_}, see `arbitrary_types_allowed` in Config')
RuntimeError: no validator found for typing.Sequence, see `arbitrary_types_allowed` in Config
```

Same behavior with or without a config file. I have no idea what exactly causes this issue, or whether it's a bug in pydantic (using `pydantic==1.5.1`). Works as expected in 3.7+ though.

**Comment by borntyping** [comment]
Thanks! I'll try and fix this on Monday :)

**Comment by borntyping** [comment]
Fixed in jsonlog-cli 6.0.1. :)

**Comment by lschmelzeisen** [comment]
Would you mind pushing to PyPI? ;)

**Comment by borntyping** [comment]
Done! (Sorry, long day.)
